src/runtime_src/core/edge/drm/zocl/edge/zocl_hwctx.c

```diff
@@ -482,3 +483,115 @@ int zocl_hw_ctx_execbuf(struct drm_zocl_dev *zdev, struct drm_zocl_hw_ctx_execbu
         ZOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(gem_obj);
     return ret;
 }
+
+static int zocl_add_hw_graph_context(struct drm_zocl_dev *zdev, struct kds_client *client,
+        struct kds_client_hw_ctx *kds_hw_ctx, struct drm_zocl_open_graph_ctx *drm_graph_ctx)
+{
+    struct zocl_hw_graph_ctx *graph_ctx = NULL;
+    int ret = 0;
+
+    list_for_each_entry(graph_ctx, &kds_hw_ctx->graph_ctx_list, link) {
+        if (graph_ctx->graph_id != drm_graph_ctx->graph_id){
+            continue;
+        }
+
+        if (graph_ctx->flags == ZOCL_CTX_EXCLUSIVE || drm_graph_ctx->flags == ZOCL_CTX_EXCLUSIVE) {
+            DRM_ERROR("%s: graph %d is already opened with exclusive context", __func__, graph_ctx->graph_id);
+            ret = -EBUSY;
+            goto out;
+        }
+
+        if (graph_ctx->flags == ZOCL_CTX_PRIMARY && drm_graph_ctx->flags != ZOCL_CTX_SHARED) {
+            DRM_ERROR("%s: graph %d is already opened with primary context", __func__, graph_ctx->graph_id);
+            ret = -EBUSY;
+            goto out;
+        }
+    }
+
+    graph_ctx = kzalloc(sizeof(*graph_ctx), GFP_KERNEL);
+    graph_ctx->flags = drm_graph_ctx->flags;
+    graph_ctx->graph_id = drm_graph_ctx->graph_id;
+    graph_ctx->hw_context = drm_graph_ctx->hw_context;
+
+    list_add_tail(&graph_ctx->link, &kds_hw_ctx->graph_ctx_list);
+    return 0;
+
+out:
+    DRM_ERROR("%s: Failed to to add the graph ctx for graph = %d", __func__, drm_graph_ctx->graph_id);
+    return ret;
+
+}
+
+int zocl_open_graph_ctx(struct drm_zocl_dev *zdev, struct drm_zocl_open_graph_ctx *drm_graph_ctx, struct drm_file *filp)
+{
+    struct kds_client_hw_ctx *kds_hw_ctx = NULL;
+    struct kds_client *client = filp->driver_priv;
+    int ret = 0;
+
+    if (!client) {
+        DRM_ERROR("%s: Invalid client", __func__);
+        return -EINVAL;
+    }
+    mutex_lock(&client->lock);
```

Comment:
As we are supporting parallel xclbin load/unload we should not lock the client. Instead of we should lock the graph list while making changes on graph list.  
When we open a graph context we should be free to work on PL xclbin.

Reply:
Hi @saifuddin-xilinx as discussed offline for now we'll have the lock on the client only. Later based on the usecase/requirement we'll update the locking mechanism.